contracts/exchange/Exchange_ERC20_1155.cairo

```diff
@@ -0,0 +1,646 @@
+        tokens_bought: Uint256,
+    ):
+end
+
+@event
+func currency_purchased(
+        caller: felt,
+        currency_bought: Uint256,
+        token_id: felt,
+        tokens_sold: Uint256,
+    ):
+end
+
+# Contract Address of ERC20 address for this swap contract
+@storage_var
+func currency_address() -> (address : felt):
+end
+
+# Contract Address of ERC1155 address for this swap contract
+@storage_var
+func token_address() -> (address : felt):
+end
+
+# Current reserves of currency
+@storage_var
+func currency_reserves(token_id : felt) -> (reserves : Uint256):
+end
+
+# Total issued LP totals
+@storage_var
+func lp_reserves(token_id : felt) -> (total : Uint256):
+end
+
+# Liquidity pool rewards in thousandths (e.g. 15 = 1.5% fee)
+@storage_var
+func lp_fee_thousands() -> (lp_fee_thousands : Uint256):
+end
+
+# Note: We use ERC1155_balanceOf(contract_address) to record token reserves
+
+@constructor
+func constructor {
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr,
+    }(
+        currency_address_: felt,
+        token_address_: felt,
+        lp_fee_thousands_: Uint256,
+    ):
+    currency_address.write(currency_address_)
+    token_address.write(token_address_)
+    lp_fee_thousands.write(lp_fee_thousands_)
+    return ()
+end
+
+#
+# Liquidity
+#
+
+@external
+func initial_liquidity {
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr,
+    }(
+        # Amount of currency supplied to LP
+        currency_amount: Uint256,
+        # ERC1155 token id
+        token_id: felt,
+        # Amount of token supplied
+        token_amount: Uint256,
+    ):
+    alloc_locals
+
+    let (caller) = get_caller_address()
+    let (contract) = get_contract_address()
+
+    let (token_address_) = token_address.read()
+    let (currency_address_) = currency_address.read()
+
+    # Only valid for first liquidity add to LP
+    let (currency_reserves_) = currency_reserves.read(token_id)
+    with_attr error_message("Only valid for initial liquidity add"):
+        assert currency_reserves_ = Uint256(0, 0)
+    end
+
+    # Transfer currency and token to exchange
+    IERC20.transferFrom(currency_address_, caller, contract, currency_amount)
+    tempvar syscall_ptr :felt* = syscall_ptr
+    IERC1155.safeTransferFrom(token_address_, caller, contract, token_id, token_amount.low)
+
+    # Assert otherwise rounding error could end up being significant on second deposit
+    let (ok) = uint256_le(Uint256(1000, 0), currency_amount)
+    with_attr error_message("Must supply larger currency for initial deposit"):
+        assert_not_zero(ok)
+    end
+
+    # Update currency reserve size for token id before transfer
+    currency_reserves.write(token_id, currency_amount)
+
+    # Initial liquidity is currency amount deposited
+    lp_reserves.write(token_id, currency_amount)
+
+    # Mint LP tokens
+    ERC1155_mint(caller, token_id, currency_amount.low)
```

Comment:
Shouldn't the third arg here be `token_amount.low`?

Reply:
No. LP tokens are issued at a rate corresponding to the currency price. (We could change this here and in the add liquidity method to use the token amount instead if you'd rather. Nifty uses the currency)